### ISSUE TITLE
Unit conversion rules: add tests and support

### ISSUE THREAD

**Issue by MichaelClerx** [ModellingWebLab/weblab-fc#156]
Choosing rule is easy*:

- if old unit not compatible with new unit
  - scan for rule that has input comp with old, output compat with new

*provided we never have to apply more than one rule

Applying:

Not sure yet!

**Comment by MichaelClerx** [comment]
Look in e.g. ICaL.txt for examples

**Comment by MichaelClerx** [comment]
```
    convert A_per_F to uA_per_cm2 by lambda rhs: rhs * chaste_membrane_capacitance
    convert uA_per_cm2 to A_per_F by lambda rhs: rhs / chaste_membrane_capacitance
    convert microamps to uA_per_cm2 by lambda rhs: (rhs / oxmeta:membrane_capacitance) * chaste_membrane_capacitance
    convert uA_per_cm2 to microamps by lambda rhs: (rhs * oxmeta:membrane_capacitance) / chaste_membrane_capacitance
```

**Comment by MichaelClerx** [comment]
The rule for inputs is:
```
x = 5mV ox:x
y = 4 + x
```
where `ox:x` denotes the oxmeta annotation `x`
becomes
```
x2 = f(5mV)  ox:x
x = f'(x2)
y = 4 + x
```
so we need both f and f' here, the documentation isn't explicit about that!

For outputs it's easier:
```
x = y ox:x
```
becomes
```
x2 = f(x) ox:x
x = y
```
but what about the time variable?

**Comment by MichaelClerx** [comment]
https://pint.readthedocs.io/en/0.11/contexts.html#defining-contexts-programmatically

**Comment by jonc125** [comment]
The documentation isn't particularly explicit because it 'just happens' as part of the general units conversion functionality in pycml. Rather than throwing an error if at any point you're asked to convert between dimensions, it checks if it has a rule that lets you do it. If you do, it applies the rule, and now you're back in the scaling-within-a-dimension case that it knows how to handle.

I think the pint context approach should deliver the same functionality, judging from the docs!

**Comment by jonc125** [comment]
Also see developer ref docs at https://pint.readthedocs.io/en/0.11/developers_reference.html#pint.Context

**Comment by jonc125** [comment]
The following works with the small patch to `units.py` shown below:
```python
import pint
import sympy as sp
from cellmlmanip.model import VariableDummy
from cellmlmanip.model import UnitStore

store = UnitStore()
microampere = store.add_unit('microampere', '1e-6 * ampere')
centimetre = store.add_unit('centimetre', '1e-2 * metre')
microfarad = store.add_unit('microfarad', '1e-6 * farad')
A_per_F = store.add_unit('A_per_F', 'ampere / farad')
uA_per_cm2 = store.add_unit('uA_per_cm2', 'microampere / (centimetre ** 2)')
uF_per_cm2 = store.add_unit('uF_per_cm2', 'microfarad / (centimetre ** 2)')

x = VariableDummy('x', uF_per_cm2)
chaste_membrane_capacitance = store.Quantity(x, x.units)

# convert A_per_F to uA_per_cm2 by lambda rhs: rhs * chaste_membrane_capacitance
f = lambda ureg, rhs: rhs * chaste_membrane_capacitance
c = pint.Context('fc')
c.add_transformation(A_per_F, uA_per_cm2, f)
store._registry.add_context(c)

q = 2 * A_per_F
q.to(uA_per_cm2, 'fc')

store._registry.enable_contexts('fc')
store.convert(q, uA_per_cm2)
store.get_conversion_factor(uA_per_cm2, A_per_F)
```

```diff --git i/cellmlmanip/units.py w/cellmlmanip/units.py
index f0055aa..027a146 100644
--- i/cellmlmanip/units.py
+++ w/cellmlmanip/units.py
@@ -363,8 +363,8 @@ class UnitStore(object):
         :returns: the magnitude of the resulting conversion factor
         """
         assert isinstance(from_unit, self._registry.Unit), 'from_unit must be a unit, not ' + str(from_unit)
-        conversion_factor = self.convert(1 * from_unit, to_unit).magnitude
-        return 1.0 if math.isclose(conversion_factor, 1.0) else conversion_factor
+        cf = self.convert(1 * from_unit, to_unit).magnitude
+        return 1.0 if isinstance(cf, numbers.Number) and math.isclose(cf, 1.0) else cf

     def _prefix_expression(self, match):
         """Takes a regex Match object from _WORD, and adds a prefix (UnitStore id), taking SI prefixes into account."""
```

**Comment by MichaelClerx** [comment]
How do we deal with variable references made in the unit conversion rules?

Can they
- be newly created variables (locals, optional with default expr, define)'?
- be updated variables (so use the new definition)?
- be variables that have an ambiguous unit themselves (so units set from RHS)?

I can only add a conversion rule once I have the symbols it contains, and at the moment we do unit conversion in a separate step long before new variables etc. are created.

Does this mean we have to start doing unit conversion inside that big loop as well? (Because the order of operations needs to be figured out as we apply the operations)

**Comment by jonc125** [comment]
According to the docs at https://chaste.cs.ox.ac.uk/trac/wiki/FunctionalCuration/ProtocolSyntax#Modelinterface references must be to an existing model variable or protocol local variable. If a reference doesn't exist the rule is ignored (with a warning). So it should be possible to check references exist at the initial units conversion phase, before other modifications.

--
Written on a smartphone; please excuse brevity!
________________________________

**Comment by MichaelClerx** [comment]
A rule like
```
    var chaste_membrane_capacitance units uF_per_cm2 = 1
    convert A_per_F to uA_per_cm2 by lambda rhs: rhs * chaste_membrane_capacitance
    convert uA_per_cm2 to A_per_F by lambda rhs: rhs / chaste_membrane_capacitance
    convert microamps to uA_per_cm2 by lambda rhs: (rhs / oxmeta:membrane_capacitance) * chaste_membrane_capacitance
    convert uA_per_cm2 to microamps by lambda rhs: (rhs * oxmeta:membrane_capacitance) / chaste_membrane_capacitance
```

 uses a new local variable though

**Comment by jonc125** [comment]
Indeed, but it's only that construct we need to check.

**Comment by MichaelClerx** [comment]
But local variables can depend on model variables, can't they?

**Comment by jonc125** [comment]
If there's a separate `define` setting up the relationship, yes. But we don't need to worry about that for the conversion rule, we only need to create the `VariableDummy` so that we can refer to it in the units-converted equation. And to create the variable all we need are its name & units, which are both given in the `var` statement.

**Comment by MichaelClerx** [comment]
And what about models that don't provide `oxmeta:membrane_capacitance`? (Because they use uA/cm^2 for currents and don't bother adding a C=1) ?

**Comment by jonc125** [comment]
Then as the docs say that rule is simply ignored (with a warning logged so the user knows that happened).

**Comment by MichaelClerx** [comment]
OK. I'll do an implementation where everything in the conversion rule has to be a simple constant then - no references

It's not at all obvious to me from the docs that the local var in a unit conversion rule can't access other variables (which would need some shuffling to make sure we evaluate in the right order), or that I couldn't use a default expression to make sure a model variable exists

Also I guess if I redefine `oxmeta:membrane_capacitance` the rule should use the old version?

**Comment by jonc125** [comment]
> OK. I'll do an implementation where everything in the conversion rule has to be a simple constant then - no references
> 
> It's not at all obvious to me from the docs that the local var in a unit conversion rule can't access other variables (which would need some shuffling to make sure we evaluate in the right order), or that I couldn't use a default expression to make sure a model variable exists

But you're not evaluating the variable when you apply the conversion rule. They're really noddy - just inserting the extra maths in the equation being converted. Evaluation doesn't happen until runtime. You don't need to care what the **definitions** of variables in the rules are, just whether they exist. So non-constant definitions are fine.

> Also I guess if I redefine `oxmeta:membrane_capacitance` the rule should use the old version?

Hmm, that could get complicated, as following the 'easy to implement' route would potentially give different results depending on the order in which unit conversions happen. I'm very tempted to say to protocol authors "don't do that"!

**Comment by MichaelClerx** [comment]
So...

- gather info on all protocol variables
- attempt to create conversion rules from known variables
  - either resolved from the model
  - or local variables. Slight issue: local variables are not created at this point, so when trying to convert the conversion rule to sympy we might need a different variable generator?)
  - protocol error if any of the model variables get redefined, or if their units are changed
- convert all units
- create local variables and redefine stuff

so when we create the lambda, we have to create our own versions of the local variables. Otherwise we have to wait until they are created in the `modify_model` loop, where the ordering of anything with dependencies is sorted out.

**Comment by jonc125** [comment]
I would say instead:

1. gather info on all protocol variables
2. for each variable reference `pvar` within each conversion rule
```
if pvar.is_local:
    if pvar.units is None:
        skip this rule with a warning
    create var and assign to pvar.model_variable
if pvar.model_variable is None:
    skip this rule with a warning
pvar.must_not_be_converted = True
```
3. add remaining conversion rules to pint
4. convert units, raising an error if asked to convert a variable with `must_not_be_converted` True
5. in `_process_protocol_variables`, don't re-create local vars created above (just add defining equation)

**Comment by MichaelClerx** [comment]
We've got this rule quite a lot:

```
var chaste_membrane_capacitance units uF_per_cm2 = 1
convert uA to uA_per_cm2 by lambda rhs: (rhs / oxmeta:membrane_capacitance) * chaste_membrane_capacitance
```
so with `X = oxmeta:membrane_capacitance` that says
```
uA / X * uF/cm^2 = uA/cm^2
```
so
```
uF/X =1 --> X = uF
```
so that means that `oxmeta:membrane_capacitance` should be in `uF`, right? 

Should we not have a rule in there that checks that it actually is? The first model I checked had it differently:

```
      <variable units="microF_per_cm2" name="C" initial_value="1" cmeta:id="membrane_capacitance">
```

**Comment by MichaelClerx** [comment]
So I could stick this in
```
    output oxmeta:membrane_capacitance units uF
```
and then if the model has it in any other units (including e.g. pF), I have to ignore this rule?

**Comment by MichaelClerx** [comment]
`ModelInterface.merge()`  doesn't do anything with unit conversion rules. Is that deliberate and should they not be merged, or is this because we hadn't got there yet?

**Comment by jonc125** [comment]
> `ModelInterface.merge()` doesn't do anything with unit conversion rules. Is that deliberate and should they not be merged, or is this because we hadn't got there yet?

It's because we hadn't got there yet. They should be merged.

**Comment by jonc125** [comment]
> ```
> var chaste_membrane_capacitance units uF_per_cm2 = 1
> convert uA to uA_per_cm2 by lambda rhs: (rhs / oxmeta:membrane_capacitance) * chaste_membrane_capacitance
> ```
> so that means that `oxmeta:membrane_capacitance` should be in `uF`, right?

I'm pretty sure it is for all models that require that particular conversion. If a model has capacitance normalised it generally has currents normalised in the same way, and either requires no conversion or a different rule.

If the rule is applied but capacitance isn't in the expected units, you'll get a dimensional inconsistency error I think (you should at some point, certainly).

**Comment by MichaelClerx** [comment]
The paci model has Cm in farad, I see. And from its scale you'd expect pF to be a more sensible unit (it's about 78pF). I'm guessing this'd be fine in the current WL because the within-equation conversion turns the pF into uF? So the rule isn't really from/to the units it specifies just from/to something similar?

**Comment by jonc125** [comment]
Exactly, it just needs to get you to the correct dimension.

**Comment by MichaelClerx** [comment]
So now if I stick in a rule demanding that C is in the units it should be in the make the equation work, then the unit conversion rule will be ignored in most cases?

**Comment by jonc125** [comment]
I'm not sure you need to put in an explicit check for that? At least if using [the approach I played with above](https://github.com/ModellingWebLab/weblab-fc/issues/156#issuecomment-614675629)?

**Comment by MichaelClerx** [comment]
Yeah I think it'd work. I just think it's pretty weird that if a user were to add a rule, saying that C has to be in uF, like it is in the equation, that'd stop a whole bunch of models working

**Comment by jonc125** [comment]
Well, if the user wanted C in uF, they'd also need to add conversion rules to get it there for cases where it's modelled differently...

**Comment by MichaelClerx** [comment]
It's not that the user wants C at all, the user just wants to convert some units, but now she has to think:

1. I'll add a rule from units to some other units
2. This rule will require a C in uF - that's certainly what I'd expect from looking at the examples we have
3. But in fact it doesn't have to be in uF, it can be in any xF
4. And if I add a rule saying "please give me C in uF", then it will stop working in many cases because we can't do unit conversion when deriving unit conversion rules

I don't think it's intuitive at all, unless you (1) expect within-equation unit conversion (which I've never seen in any other tool, certainly not in matlab which is where our users will come from), and (2) have some knowledge of how the unit conversion in FC works

**Comment by MichaelClerx** [comment]
> Well, if the user wanted C in uF, they'd also need to add conversion rules to get it there for cases where it's modelled differently...

I also think we should update all those models where "1 uF/cm^2" is annotated as "membrane_capacitance". It really isn't. Those models simply don't specify a capacitance (although you might be able to work one out if they specify a membrane area).

**Comment by jonc125** [comment]
> I also think we should update all those models where "1 uF/cm^2" is annotated as "membrane_capacitance". It really isn't. Those models simply don't specify a capacitance (although you might be able to work one out if they specify a membrane area).

Yeah, this is something I've discussed with @mirams before. We should document what the expectations are for some of our ontology terms so people use them properly! And possibly add terms like "normalised_membrane_capacitance". A similar issue occurs with currents, where polarity matters.

**Comment by jonc125** [comment]
> It's not that the user wants C at all, the user just wants to convert some units, but now she has to think:
> ...
> I don't think it's intuitive at all, unless you (1) expect within-equation unit conversion (which I've never seen in any other tool, certainly not in matlab which is where our users will come from), and (2) have some knowledge of how the unit conversion in FC works.

I think this more says we need to document how units conversion rules are intended to behave, which I think we do in the language docs, but possibly could be improved (and also in the tutorial protocols?).

**Comment by MichaelClerx** [comment]
Closed in #179
